staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -338,6 +365,33 @@ func New(opts Options) (*Authenticator, error) {
 	return authenticator, nil
 }
 
+// discoveryURLRoundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type discoveryURLRoundTripper struct {
+	base         http.RoundTripper
+	discoveryURL *url.URL
+	// urlToRewrite is the URL to rewrite to
+	// https://<discoveryURL>/.well-known/openid-configuration
+	urlToRewrite string
+}
+
+func (t *discoveryURLRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && req.URL.String() == t.urlToRewrite {
+		clone := req.Clone(req.Context())
+		clone.Host = ""
+		clone.URL.RawPath = ""
```

Comment:
if `t.discoveryURL` is fully constructed at startup, can we just do `clone.URL = t.discoveryURL` and override everything?

Reply:
This doesn't seem to work `clone.URL = t.discoveryURL`. Doing this results in `path` missing the leading slash.

```
before: &url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"auth.example.com", Path:"/.well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
after: &url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"127.0.0.1:52528", Path:".well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
```

Reply:
I think it would work with or without the leading slash (the stringification adds it)

Reply:
that's what I thought but the unit tests fail with `400 Bad Request` and I don't see the request come into the tests claim server with

```go
discoveryURL = discoveryURL.JoinPath(wellKnownEndpointPath)
clone.URL = t.discoveryURL
```

This works as expected and I've updated the PR with it
```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

Reply:
_Debugging `discoveryURL = discoveryURL.JoinPath(wellKnownEndpointPath)` further_

Dumping the request with
```go
	bytes, _ := httputil.DumpRequestOut(clone, true)
```

I see the path is without the leading slash
```bash
GET .well-known/openid-configuration HTTP/1.1
Host: 127.0.0.1:56692
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip
```

Reply:
that's super-weird... I'm surprised the go round tripper is happy writing an invalid request like that